src/huggingface_hub/utils/common.py

```diff
@@ -0,0 +1,34 @@
+from multiprocessing.sharedctypes import Value
+from typing import Any, Iterable, Iterator, List
+
+
+def chunk_iterable(
+    iterable: Iterable[Any],
+    chunk_size: int,
+) -> Iterator[List[Any]]:
+    """
+    Returns an iterator over iterable in chunks of size chunk_size.
+
+    ```python
+    >>> iterable = range(128)
+    >>> chunked_iterable = chunk_iterable(iterable, chunk_size=8)
+    >>> next(chunked_iterable)
+    # [0, 1, 2, 3, 4, 5, 6, 7]
+    >>> next(chunked_iterable)
+    # [8, 9, 10, 11, 12, 13, 14, 15]
+    ```
+    """
+
+    def _chunk_iter(iter: Iterable[Any]) -> Iterator[List[Any]]:
+        chunk = []
+        for elem in iterable:
+            chunk.append(elem)
+            if len(chunk) >= chunk_size:
+                yield chunk
+                chunk = []
+        if len(chunk):
+            yield chunk
```

Comment:
Can we use https://stackoverflow.com/a/312464/835629 ?

```python
def chunks(lst, n):
    """Yield successive n-sized chunks from lst."""
    for i in xrange(0, len(lst), n):
        yield lst[i:i + n]
```

Reply:
^ this solution does not work with a general `Iterable` as input (`lst` needs to support indexing)

No strong opinion though

Reply:
975568f proposes an alternative version, closer to the suggested solution

Reply:
I'll let more python-oriented folks comment

Reply:
Since the question arose, I would say an even cleaner solution would be to avoid building lists for each chunk but rather return iterators as well. This topic is highly discussed on stackoverflow but a solution that I found convincing is this one https://stackoverflow.com/a/8998040:

```python
import itertools
from typing import Iterable, TypeVar


T = TypeVar("T")


def chunk_iterable(iterable: Iterable[T], chunk_size: int) -> Iterable[Iterable[T]]:
    """Iterate an iterable by chunks of size `chunk_size`.
    
    The last chunk can be smaller than `chunk_size`.

    Taken from https://stackoverflow.com/a/8998040.
    """
    if not isinstance(chunk_size, int) or chunk_size <= 0:
        raise ValueError("`chunk_size` must be a strictly positive integer (>0).")

    iterator = iter(iterable)
    while True:
        try:
            next_item = next(iterator)
        except StopIteration:
            return
        yield itertools.chain((next_item,), itertools.islice(iterator, chunk_size - 1))
```

The two main benefits I see are:
1. items are consumed from the iterable when they are actually used (except the first item of each chunk :/ ) -> might be useful in case of a generator
2. lists are not created -> might be useful in case of large chunk_size

=> benefits are quite small apart in some corner cases

**Note:** if you go for this solution, you'll need to adapt the tests since they expect chunks as List.

Reply:
Quick note about the `TypeVar`: it is just a way to tell that the item type in the output will be the same as the one in the input. So for example here, mypy would know that an iterable of `str` will become an iterable of chunks of `str`.

More about it: https://docs.python.org/3/library/typing.html#typing.TypeVar